pkg/traceql/engine_metrics_average.go

```diff
@@ -0,0 +1,540 @@
+			lookups[i] = []Attribute{
+				NewScopedAttribute(AttributeScopeSpan, false, attr.Name),
+				NewScopedAttribute(AttributeScopeResource, false, attr.Name),
+			}
+		} else {
+			lookups[i] = []Attribute{attr}
+		}
+	}
+
+	aggNum := len(lookups)
+
+	switch aggNum {
+	case 2:
+		return newAvgAggregator[FastStatic2, StaticVals2](attr, by, lookups, start, end, step)
+	case 3:
+		return newAvgAggregator[FastStatic3, StaticVals3](attr, by, lookups, start, end, step)
+	case 4:
+		return newAvgAggregator[FastStatic4, StaticVals4](attr, by, lookups, start, end, step)
+	case 5:
+		return newAvgAggregator[FastStatic5, StaticVals5](attr, by, lookups, start, end, step)
+	default:
+		return newAvgAggregator[FastStatic1, StaticVals1](attr, by, lookups, start, end, step)
+	}
+}
+
+func newAvgAggregator[F FastStatic, S StaticVals](attr Attribute, by []Attribute, lookups [][]Attribute, start, end, step uint64) SpanAggregator {
+	var fn func(s Span) float64
+
+	switch attr {
+	case IntrinsicDurationAttribute:
+		fn = func(s Span) float64 {
+			return float64(s.DurationNanos()) / float64(time.Second)
+		}
+	default:
+		fn = func(s Span) float64 {
+			f, a := FloatizeAttribute(s, attr)
+			if a == TypeNil {
+				return math.Float64frombits(normalNaN)
+			}
+			return f
+		}
+	}
+
+	return &AvgOverTimeSpanAggregator[F, S]{
+		series:          map[F]AvgOverTimeSeries[S]{},
+		getSpanAttValue: fn,
+		by:              by,
+		byLookups:       lookups,
+		start:           start,
+		end:             end,
+		step:            step,
+	}
+}
+
+func (g *AvgOverTimeSpanAggregator[F, S]) Observe(span Span) {
+	if !g.getGroupingValues(span) {
+		return
+	}
+
+	s := g.getSeries()
+	interval := IntervalOf(span.StartTimeUnixNanos(), g.start, g.end, g.step)
+	if interval == -1 {
+		return
+	}
+	inc := g.getSpanAttValue(span)
+	if math.IsNaN(inc) {
+		return
+	}
+	s.count[interval]++
+	mean, c := averageInc(s.avg[interval], inc, s.count[interval], s.compensation[interval])
+	s.avg[interval] = mean
+	s.compensation[interval] = c
+}
+
+func averageInc(mean, inc, count, compensation float64) (float64, float64) {
+	if math.IsNaN(mean) && !math.IsNaN(inc) {
+		// When we have a proper value in the span we need to initialize to 0
+		mean = 0
+	}
+	if math.IsInf(mean, 0) {
+		if math.IsInf(inc, 0) && (mean > 0) == (inc > 0) {
+			// The `current.val` and `new` values are `Inf` of the same sign.  They
+			// can't be subtracted, but the value of `current.val` is correct
+			// already.
+			return mean, compensation
+		}
+		if !math.IsInf(inc, 0) && !math.IsNaN(inc) {
+			// At this stage, the current.val is an infinite. If the added
+			// value is neither an Inf or a Nan, we can keep that mean
+			// value.
+			// This is required because our calculation below removes
+			// the mean value, which would look like Inf += x - Inf and
+			// end up as a NaN.
+			return mean, compensation
+		}
+	}
+	mean, c := kahanSumInc(inc/count-mean/count, mean, compensation)
```

Comment:
This is a neat algorithm.  Just checking my understanding - both inc and mean are intended to be divided by the new count, correct? It seems right.

Reply:
Correct!